### ISSUE TITLE
Apply stricter rules on return type in Object.assign

### ISSUE THREAD

**Issue by ParkAsher** [microsoft/TypeScript#52511]
I defined enum called "Status".
```typescript
enum Status {
    Initialized = 'Initialized',
    Pending = 'Pending',
    Complete = 'Complete',
}
```
And, i defined return type called "StatusObject".
```typescript
type StatusObject = {
    [key in Status]?: keyof typeof Status;
};
```
I want to get object type result like this.
```typescript
{
      Initialized: "initialized",
      Pending: "pending",
      Complete: "complete",
}
```
So, i declared a function  called "getStatusObject" which return object that return type is StatusObject
```
function getStatusObject(): StatusObject {
    return Object.keys(Status).reduce((arr, status) => {
        return Object.assign(arr, { [status]: status.toLowerCase() });
    }, {} as StatusObject);
}
```

I thought the function must be giving an error.
Because, the function expected to return "StatusObject" type.
In "StatusObject", key is Status type and value is string literal union type.
But, after Obejct.reduce fucntion, [x: string] : string properties in object arr.
So, expected return type and return value do not match.

But, don't report type error.

**Comment by jcalz** [comment]
If you want this issue to be addressed by the TS team you should go back and follow the instructions in the new issue template.  Otherwise you risk having this closed and getting a message like [this one](https://github.com/microsoft/TypeScript/issues/51728#issuecomment-1335574919)

> To expedite the triage process, we need everyone to follow the issue template and instructions.
> 
> When you clicked "Create New Issue", the issue form was pre-populated with a template and some instructions. We need you to read those instructions completely, follow them, and then fill in all the fields in that template.
> 
> We are not able to assist with issues that don't follow the template instructions as they represent a significantly larger amount of work compared to issues which are correctly specified. Thank you for understanding.
